### PR TITLE
Support unit tests with realm in Flutter projects

### LINKED ISSUE #791: App Flutter tests fail to run in VSCode
[See PR comments: https://github.com/realm/realm-dart/pull/660](https://github.com/realm/realm-dart/pull/660)

### LINKED ISSUE #870: unit tests using the realm
Hello everyone, I hope you are all well!

I would like to know if there is any way to integrate the realm with the unit testing features with mockito/mocktail.
I'm trying to test the realm mapping class, but I'm not having success.

Is there any way to mock the return results?

Thanks!

### PR DESCRIPTION
**Pull request by nielsenko**
- Avoid opening plugin lib when running `flutter test`
- Allow running `dart run realm install` in flutter project to install realm lib for unit tests
- Use new --flavor flag

Fixes #791, fixes #870

### CODE REVIEW

File: lib/src/cli/common/target_os_type.dart [hunk 1]
```diff
@@ -26,8 +26,24 @@ enum TargetOsType {
   windows,
 }
 
+// Cannot use Dart 2.17 enhanced enums, due to stupid issue with build_cli :-/
```
Comment (desistefanova) [inline lib/src/cli/common/target_os_type.dart:29]: Is there any nice way :) to explain this "stupid issue with build_cli :-/"?

**Review by desistefanova — COMMENTED** [review]
Is there any simple steps that I can try this on my machine?

File: lib/src/cli/common/target_os_type.dart (hunk 1 above)
Comment (nielsenko) [reply lib/src/cli/common/target_os_type.dart:29]: Problem is that `build_cli` will add the ctor as an option as well

**Comment by nielsenko** [comment]
> Is there any simple steps that I can try this on my machine?

You can make a flutter project with tests that use realm. When running `flutter test` you will get a dlopen error. Then run `flutter pub run realm install`, which will download the desktop lib, and last re-run `flutter test`. Second time around it should work.. or at least it should load the lib correctly 😄

File: lib/src/cli/install/install_command.dart [hunk 2]
```diff
@@ -205,36 +194,24 @@ class InstallCommand extends Command<void> {
...
+        abort("Invalid target OS: null.");
+      }
     } else {
-      throw Exception("Unsuported platform ${Platform.operatingSystem}");
+      options.targetOsType ??= getTargetOS();
+      if ((options.targetOsType == TargetOsType.ios || options.targetOsType == TargetOsType.android) && isDart) {
+        throw Exception("Invalid package name $packageName for target OS ${options.targetOsType}");
+      }
     }
   }
```
Comment (desistefanova) [inline lib/src/cli/install/install_command.dart:211]: Maybe this could be something like this:
``` dart
void validateOptions() {
    switch (options.flavor) {
      case Flavor.flutter:
        if (options.targetOsType == null) {
          abort("Invalid target OS: null.");
        }
        break;
      case Flavor.dart:
        options.targetOsType ??= getTargetOS();
        if ((options.targetOsType == TargetOsType.ios || options.targetOsType == TargetOsType.android)) {
          throw Exception("Invalid package name $packageName for target OS ${options.targetOsType}");
        }
        break;
    }
  }
  ```

Comment (desistefanova) [inline lib/src/cli/install/install_command.dart:211]: What about something like:
```dart
void validateOptions() {
    switch (options.flavor) {
      case Flavor.flutter:
        if (options.targetOsType == null) {
          abort("Invalid target OS: null.");
        }
        break;
      case Flavor.dart:
        options.targetOsType ??= getTargetOS();
        if ((options.targetOsType == TargetOsType.ios || options.targetOsType == TargetOsType.android)) {
          throw Exception("Invalid package name $packageName for target OS ${options.targetOsType}");
        }
        break;
    }
  }
```

**Review by desistefanova — COMMENTED** [review]
I think we could have solved the problem with fewer changes. I'm afraid we might break something else by introducing this `Flavor` enum.

**Comment by desistefanova** [comment]
@nielsenko Is it possible to have commands like these:
`dart run realm_dart install` and `flutter pub run realm install`
instead of
`dart realm install --flavor dart`  or   `flutter pub run realm install --flavor dart`.
It's quite confusing to select flavor dart when you are installing flutter.
It will be better if we can ship this fix without any changes at the existing commands, just allowing the new command `flutter pub run realm install` to be run, because we don't have a way to test these commands before to release.

**Comment by nielsenko** [comment]
> @nielsenko Is it possible to have commands like these: `dart run realm_dart install` and `flutter pub run realm install` instead of `dart realm install --flavor dart` or `flutter pub run realm install --flavor dart`. It's quite confusing to select flavor dart when you are installing flutter. It will be better if we can ship this fix without any changes at the existing commands, just allowing the new command `flutter pub run realm install` to be run, because we don't have a way to test these commands before to release.

As a user you don't have to add the flavor flag. Both of the commands you mention should suffice, as the default does the right thing.

**Review by blagoev — CHANGES_REQUESTED** [review]

File: lib/src/cli/common/target_os_type.dart (hunk 1 above)
Comment (nielsenko) [reply lib/src/cli/common/target_os_type.dart]: I have dropped _stupid_ for _an_.

File: CHANGELOG.md [hunk 3]
```diff
@@ -11,6 +11,8 @@
 
 ### Fixed
 * Support mapping into `SyncSessionErrorCode` for "Compensating write" with error code 231. ([#1022](https://github.com/realm/realm-dart/pull/1022))
+* Previously it was difficult to setup a flutter project with realm, so that `flutter test` would work
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply CHANGELOG.md:23]: I can change it like that, but it will not be shorter :-)

File: lib/src/cli/install/options.dart [hunk 4]
```diff
@@ -23,19 +23,17 @@ part 'options.g.dart';
...
 class Options {
-  @CliOption(help: "Required for Flutter. The target OS to install binaries for.", abbr: "t")
+  @CliOption(help: 'The target OS to install binaries for.', abbr: 't')
   TargetOsType? targetOsType;
 
-  // packageName defaults to `realm_dart` since "realm" is always set by the build scripts.
-  // The `Install` command when used by end users will not require the package name argument
-  @CliOption(defaultsTo: 'realm_dart', help: "Optional. The realm package name to install binaries for.", abbr: "p", allowed: ['realm', 'realm_dart'])
-  final String? packageName;
+  @CliOption(help: 'The flavor to install binaries for.', abbr: 'f', defaultsTo: Flavor.dart)
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/cli/install/options.dart:29]: It does

File: lib/src/cli/install/install_command.dart [hunk 5]
```diff
@@ -174,27 +167,23 @@ class InstallCommand extends Command<void> {
...
       var realmPubspecFile = await File(path).readAsString();
       final pubspec = Pubspec.parse(realmPubspecFile);
-      if (pubspec.name != packageName) {
-        throw Exception("Unexpected package name `${pubspec.name}` at $path. Realm install command expected package `$packageName`");
-      }
-
       return pubspec;
-    } on Exception catch (e) {
-      throw Exception("Error parsing package `$packageName` pubspec at $path. Error $e");
+    } catch (e) {
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/cli/install/install_command.dart:175]: Or perhaps just remove the catch altogether?

File: lib/src/native/realm_core.dart [hunk 6]
```diff
@@ -2004,53 +2004,54 @@ class _RealmCore {
...
-      return path.join(getFilesPath(), _appDir);
-    } else if (Platform.isWindows) {
-      return _appDir ?? Directory.current.absolute.path;
-    } else if (Platform.isLinux) {
-      String appSupportDir =
-          PlatformEx.fromEnvironment("XDG_DATA_HOME", defaultValue: PlatformEx.fromEnvironment("HOME", defaultValue: Directory.current.absolute.path));
-      return path.join(appSupportDir, ".local/share", _appDir);
-    } else if (Platform.isMacOS) {
-      return _appDir ?? Directory.current.absolute.path;
+      return getFilesPath();
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/native/realm_core.dart:2066]: An an Android and iOS app already uses a directory local to the app. No need to make a subdirectory in it.. but you are are right this can break stuff for existing clients. Should we mention it under breaking changes, or revert a live with the odd extra directory level?

File: lib/src/native/realm_core.dart [hunk 7]
```diff
@@ -2004,53 +2004,54 @@ class _RealmCore {
...
+        : Platform.isMacOS
+            ? 'realm.framework/realm' // use catalyst
+            : Platform.isLinux
+                ? "librealm_plugin.so"
+                : throw UnsupportedError("Platform ${Platform.operatingSystem} is not supported");
+
+    final pluginLib = DynamicLibrary.open(plugin);
+    final getAppDirFunc = pluginLib.lookupFunction<Pointer<Int8> Function(), Pointer<Int8> Function()>("realm_dart_get_app_directory");
+    final dirNamePtr = getAppDirFunc();
+    final dirName = Platform.isWindows ? dirNamePtr.cast<Utf16>().toDartString() : dirNamePtr.cast<Utf8>().toDartString();
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/native/realm_core.dart:2052]: arh.. you mean `dirNamePtr`

File: lib/src/cli/install/install_command.dart [hunk 8]
```diff
@@ -205,36 +194,24 @@ class InstallCommand extends Command<void> {
     await downloadAndExtractBinaries(binaryPath, realmPubspec, archiveName);
 
     print("Realm install command finished.");
+    exit(0);
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/cli/install/install_command.dart:197]: `downloadAndExtractBinaries` is to blame

[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/cli/install/install_command.dart:197]: I think what happens is that there are un-awaited futures that are still around, maybe for handling timeout. Unless we explicitly exit the vm will wait for those before exiting.

File: CHANGELOG.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (blagoev) [reply CHANGELOG.md:23]: True, let me try again :)

Support `realm pub run install` command can be executed in Flutter projects which need to run Flutter unit tests and widget tests. This will make sure the required native binary is at the correct location.

```suggestion
Support install command in flutter projects that use unit and widget tests. 
```

File: lib/src/cli/install/install_command.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (blagoev) [reply lib/src/cli/install/install_command.dart:175]: well we want to print the path to the pubspec. Also this guards us against cryptic errors

File: lib/src/cli/install/install_command.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (blagoev) [reply lib/src/cli/install/install_command.dart:197]: sure, got it. maybe we can add a comment about that above that line. Its not obvious where these future are. But we don't care if the command has finished alrady.

File: lib/src/native/realm_core.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (blagoev) [reply lib/src/native/realm_core.dart:2052]: yes, sorry.

File: lib/src/cli/install/install_command.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/cli/install/install_command.dart:197]: .. found it. In `downloadAndExtractBinaries` you create two `HttpClient` one that is never used , but correctly closed, and one that is used, but not closed eagerly.

File: lib/src/cli/install/install_command.dart [hunk 9]
```diff
@@ -161,7 +154,7 @@ class InstallCommand extends Command<void> {
           "Run the 'dart run $packageName install` command from the root directory of your application");
     }
 
-    final realmDartPackage = packageConfig.packages.where((p) => p.name == packageName).firstOrNull;
+    final realmDartPackage = packageConfig.packages.where((p) => p.name == Flavor.flutter.packageName || p.name == Flavor.dart.packageName).firstOrNull;
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/cli/install/install_command.dart]: Fixed.. it involves a few more changes

File: lib/src/cli/install/install_command.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/cli/install/install_command.dart]: Revert to `on Exeception`

File: lib/src/cli/install/options.dart [hunk 10]
```diff
@@ -23,19 +23,17 @@ part 'options.g.dart';
...
+  Flavor flavor;
 
-  @CliOption(hide: true, defaultsTo: false)
-  //use to debug install command
-  bool? debug;
-  
-  Options({this.targetOsType, this.packageName, this.debug});
+  // use to debug install command
+  @CliOption(hide: true)
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/cli/install/options.dart]: added `defaultsTo`

File: lib/src/native/realm_core.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/native/realm_core.dart:2065]: This diff looks odd..

[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/native/realm_core.dart:2065]: `_appDir`  was always empty for iOS and Android the way it was calculated before. Files created on Android or iOS are local to the app by default.

File: lib/src/native/realm_core.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/native/realm_core.dart:2052]: In that case `.toDartString()` will throw `UnsupportedError`
```dart
extension Utf8Pointer on Pointer<Utf8> {
  ...
  String toDartString({int? length}) {
    _ensureNotNullptr('toDartString');
    ...
  }
  ...
    void _ensureNotNullptr(String operation) {
    if (this == nullptr) {
      throw UnsupportedError(
          "Operation '$operation' not allowed on a 'nullptr'.");
    }
  }
```
I assume that is fine.. this is a bug on us, if it happens

File: lib/src/cli/install/install_command.dart (hunk 2 above)
Comment (nielsenko) [reply lib/src/cli/install/install_command.dart:208]: Using switch on just two options like this feels a little odd to me.

File: lib/src/native/realm_core.dart [hunk 11]
```diff
@@ -1997,53 +1997,54 @@ class _RealmCore {
...
+      return path.join(appSupportDir, ".local/share", _getAppDirectoryFromPlugin());
+    }
+
+    if (Platform.isMacOS) {
+      return _getAppDirectoryFromPlugin();
+    }
+
+    if (Platform.isWindows) {
+      return _getAppDirectoryFromPlugin();
     }
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/native/realm_core.dart:2085]: There are not more calls. Only one branch is taken. And I think it reads better as well.

File: lib/src/cli/install/install_command.dart [hunk 12]
```diff
@@ -101,16 +96,14 @@ class InstallCommand extends Command<void> {
...
     final versionsFile = File(path.join(binariesPath, versionFileName));
-
-    if (!await versionsFile.exists()) {
-      return false;
-    }
-
-    final existingVersion = await versionsFile.readAsString();
-    if (expectedVersion == existingVersion) {
-      print("Realm binaries for $packageName@$expectedVersion already downloaded");
+    if (await versionsFile.exists()) {
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/cli/install/install_command.dart:99]: one less comparison wins, but in general I prefer early exit in most cases.

File: lib/src/native/realm_core.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (blagoev) [reply lib/src/native/realm_core.dart:2065]: Could you double check that. Why would it be empty for iOS and Android. It because of these platforms that `realm_dart_get_app_directory` was introduced.

File: lib/src/native/realm_core.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (blagoev) [reply lib/src/native/realm_core.dart:2052]: I think we are changing too much here. Previously we did handle this gracefully by appending an empty string

File: lib/src/cli/install/install_command.dart [hunk 13]
```diff
@@ -85,13 +78,15 @@ class InstallCommand extends Command<void> {
...
+
+    if (Flavor.values.map((f) => f.packageName).contains(projectPubspec.name)) {
+      print(//
+          'Running install command inside ${projectPubspec.name} package which is the development package for Realm.\n'
           'Skipping download as it is expected that you build the packages manually.');
       return true;
     }
 
-    if (realmPubspec.publishTo == 'none') {
+    if (realmPubspec.publishTo == 'none' && !debug) {
```
Comment (desistefanova) [inline lib/src/cli/install/install_command.dart:89]: Do we really need this check? What if we remove it in order to be able to test with path dependency before the release?

**Review by desistefanova — COMMENTED** [review]
I think it is ready. I have only some minor questions and suggestions.

File: lib/src/cli/install/install_command.dart [hunk 14]
```diff
@@ -208,33 +195,20 @@ class InstallCommand extends Command<void> {
...
-    if (Platform.isWindows) {
-      return TargetOsType.windows;
-    } else if (Platform.isLinux) {
-      return TargetOsType.linux;
-    } else if (Platform.isMacOS) {
-      return TargetOsType.macos;
+    if (isFlutter) {
+      if (options.targetOsType == null) {
+        abort("Invalid target OS: null.");
+      }
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/cli/install/install_command.dart:201]: You cannot run plain dart applications on Android and iOS

File: lib/src/cli/install/install_command.dart (hunk 13 above)
Comment (nielsenko) [reply lib/src/cli/install/install_command.dart:89]: We already have the debug flag. Why not use it here?

File: lib/src/cli/install/install_command.dart [hunk 15]
```diff
@@ -208,33 +195,20 @@ class InstallCommand extends Command<void> {
...
-      throw Exception("Unsupported platform ${Platform.operatingSystem}");
+      options.targetOsType ??= getTargetOS();
+      if ((options.targetOsType == TargetOsType.ios || options.targetOsType == TargetOsType.android) && isDart) {
+        throw Exception("Invalid package name $packageName for target OS ${options.targetOsType}");
+      }
     }
   }
 
+  TargetOsType getTargetOS() => Platform.operatingSystem.asTargetOsType ?? (throw UnsupportedError("Unsupported platform ${Platform.operatingSystem}"));
+
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/cli/install/install_command.dart:211]: `asTargetOsType` will return `null` in that case, so it will throw

File: lib/src/native/realm_core.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (nielsenko) [reply lib/src/native/realm_core.dart:2090]: There is a difference between:

- realm_dart_get_app_directory, and
- realm_dart_get_files_path

The first is implemented for linux, macos, and windows, the later for android and ios.

**Review by blagoev — APPROVED** [review]
We discussed to error out with a better error message. 

If for example on Windows `SHGetKnownFolderPath` fails temporary we don't want to create a realm in the current directory and then later create a new realm in the correct place effectively ignoring the first realm which might have already populated data.
